File: Cloud Mail Samples/Gmail/c++/gmail.cpp
/*
 * Cloud Mail 2024 C++ Edition - Sample Project
 *
 * This sample project demonstrates the usage of Cloud Mail in a 
 * simple, straightforward way. It is not intended to be a complete 
 * application. Error handling and other checks are simplified for clarity.
 *
 * www.nsoftware.com/cloudmail
 *
 * This code is subject to the terms and conditions specified in the 
 * corresponding product license agreement which outlines the authorized 
 * usage and restrictions.
 */

#include <stdio.h>
#include <ctype.h>
#include <string.h>
#include <stdlib.h>
#include "../../include/cloudmail.h"

#define LINE_LEN 80

class MyGMAIL : public Gmail
{
	bool isListingEmails;
public:

	MyGMAIL()
	{
		isListingEmails = false;
	}

	virtual int FireMessageInfo(GmailMessageInfoEventParams* e)
	{
		if (isListingEmails) {
			printf("%s\n", e->Id);
		}
		return 0;
	}

	virtual int FireSSLServerAuthentication(GmailSSLServerAuthenticationEventParams* e)
	{
		if (e->Accept) return 0;
		printf("Server provided the following certificate:\nIssuer: %s\nSubject: %s\n",
			e->CertIssuer, e->CertSubject);
		printf("The following problems have been determined for this certificate: %s\n", e->Status);
		printf("Would you like to continue anyways? [y/n] ");
		if (getchar() == 'y') e->Accept = true;
		else exit(0);
		return 0;
	}
	
	void setIsListingEmais(bool value) {
		isListingEmails = value;
	}

};

int main(int argc, char * argv[])
{
		MyGMAIL gmail;

		char command[LINE_LEN];     // user's command
		char buffer[LINE_LEN];      // text buffer
		char *argument;             // arguments following command
		char *currFolder = "";

		int msgnum = 0;             // current message number for next command
		char msg_range[LINE_LEN];
		char msg_limit[10];
		int ret_code;
	
		gmail.SetOAuthServerAuthURL("https://accounts.google.com/o/oauth2/auth");
		gmail.SetOAuthServerTokenURL("https://accounts.google.com/o/oauth2/token");
		printf("Enter your OAuth Client ID: ");
		fgets(command, LINE_LEN, stdin);
		if (strlen(command) > 0 && command[strlen(command) - 1] == '\n') {
				command[strlen(command) - 1] = '\0';
		}
		gmail.SetOAuthClientId(command);
		printf("Enter your OAuth client secret: ");
		fgets(command, LINE_LEN, stdin);
		if (strlen(command) > 0 && command[strlen(command) - 1] == '\n') {
				command[strlen(command) - 1] = '\0';
		}
		gmail.SetOAuthClientSecret(command);
		printf("Enter desired authorization scopes (leave a space between each one if providing multiple): ");
		fgets(command, LINE_LEN, stdin);
		if (strlen(command) > 0 && command[strlen(command) - 1] == '\n') {
				command[strlen(command) - 1] = '\0';
		}
		gmail.SetOAuthAuthorizationScope(command);

		ret_code = gmail.Authorize();

		if (ret_code) goto done;

		printf("\nAuthorization Successful\n");

		printf("Type \"?\" for a list of commands.\n\n");
		while (1)
		{
			printf("> ");
			fgets(command, LINE_LEN, stdin);
			buffer[strlen(command) - 1] = '\0';
			argument = strtok(command, " \n\t");

			if (!strcmp(command, "?") || !strcmp(command, ""))
			{

				printf("Gmail Commands\n");
				printf("l                   list next 100 messages\n");
				printf("v <message id>      view the content of a message id\n");
				printf("c                   compose and send a message\n");
				printf("?                   display options\n");
				printf("q                   quit\n");

			}

			else if (!strcmp(command, "l"))
			{
				gmail.setIsListingEmais(true);
				gmail.Config("ResponseType=4");
				ret_code = gmail.ListMessages("", "");
			}

			else if (!strcmp(command, "v"))
			{
				argument = strtok(NULL, " \t\n");
				int messageNum = atoi(argument);
				gmail.setIsListingEmais(false);
				gmail.Config("ResponseType=1");
				ret_code = gmail.RetrieveMessageRaw(argument);

				char* theMessage;
				int messageLength;

				gmail.GetMessage(theMessage, messageLength);
				printf("%s\n", theMessage);
			}

			else if (!strcmp(command, "c"))
			{
				printf("Send To: ");
				char sendTo[LINE_LEN];
				fgets(sendTo, LINE_LEN, stdin);
				sendTo[strcspn(sendTo, "\n")] = 0;
				gmail.SetMessageTo(sendTo);

				printf("cc: ");
				char cc[LINE_LEN];
				fgets(cc, LINE_LEN, stdin);
				cc[strcspn(cc, "\n")] = 0;
				gmail.SetMessageCc(cc);

				printf("bcc: ");
				char bcc[LINE_LEN];
				fgets(bcc, LINE_LEN, stdin);
				bcc[strcspn(bcc, "\n")] = 0;
				gmail.SetMessageBcc(bcc);

				printf("Subject: ");
				char subject[LINE_LEN];
				fgets(subject, LINE_LEN, stdin);
				subject[strcspn(subject, "\n")] = 0;
				gmail.SetMessageSubject(subject);

				printf("Message Body Type(text or html): ");
				char messageBodyType[LINE_LEN];
				fgets(messageBodyType, LINE_LEN, stdin);
				messageBodyType[strcspn(messageBodyType, "\n")] = 0;
         		while (strcmp(messageBodyType, "text") != 0 && strcmp(messageBodyType, "html") != 0) {
					printf("Type either text or html: ");
					fgets(messageBodyType, LINE_LEN, stdin);
					messageBodyType[strcspn(messageBodyType, "\n")] = 0;
				}
				gmail.SetMessageBodyContentType(messageBodyType);

				printf("Message Content: ");
				char messageContent[LINE_LEN];
				fgets(messageContent, LINE_LEN, stdin);
				messageContent[strcspn(messageContent, "\n")] = 0;
				gmail.SetMessageBodyContent(messageContent);

				printf("Attachments(semicolon separated): ");
				char attachments[LINE_LEN];
				fgets(attachments, LINE_LEN, stdin);
				gmail.SetMessageAttachments(attachments);
				ret_code = gmail.SendMail();

				if (!ret_code) {
					printf("Message sent!\n");
				}

			}

			else if (!strcmp(command, "q"))
			{
				exit(0);
			}

			else
			{
				printf("Bad command / Not implemented in demo.\n");
			} // end of command checking

			ret_code = gmail.GetLastErrorCode();

			if (ret_code)     // Got an error.  The user is done.
			{
				printf("\nError: %d", ret_code);
				if (gmail.GetLastError())
				{
					printf(" \"%s\"\n", gmail.GetLastError());
				}
			}
			ret_code = 0;   // flush out error
		}  // end of main while loop

	done:
		if (ret_code)     // Got an error.  The user is done.
		{
			printf("\nError: %d", ret_code);
			if (gmail.GetLastError())
			{
				printf(" \"%s\"\n", gmail.GetLastError());
			}
		}
		printf("Exiting... (press enter)\n");
		getchar();
		exit(ret_code);
		return 0;
}

File: Cloud Mail Samples/Gmail/javascript/gmail.js
/*
 * Cloud Mail 2024 JavaScript Edition - Sample Project
 *
 * This sample project demonstrates the usage of Cloud Mail in a 
 * simple, straightforward way. It is not intended to be a complete 
 * application. Error handling and other checks are simplified for clarity.
 *
 * www.nsoftware.com/cloudmail
 *
 * This code is subject to the terms and conditions specified in the 
 * corresponding product license agreement which outlines the authorized 
 * usage and restrictions.
 */
 
const readline = require("readline");
const cloudmail = require("@nsoftware/cloudmail");

if(!cloudmail) {
  console.error("Cannot find cloudmail.");
  process.exit(1);
}
let rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout
});

let gmail1 = new cloudmail.gmail();
let isListingEmails = false;

main();

async function main() {
  // Set up event handlers.
  gmail1.on("MessageInfo", (e) => {
    if (isListingEmails)
      console.log(e.id);
  });

  // Authorize the client.
  gmail1.getOAuth().setClientId("922417066392-f9fomlr35hffi9gkq5te61nr47rtk00q.apps.googleusercontent.com");
  gmail1.getOAuth().setClientSecret("o8tCrbErRHjAh42whuvXGW0R");
  gmail1.getOAuth().setServerAuthURL("https://accounts.google.com/o/oauth2/auth");
  gmail1.getOAuth().setServerTokenURL("https://accounts.google.com/o/oauth2/token");
  gmail1.getOAuth().setAuthorizationScope("https://mail.google.com/");
  gmail1.authorize();

  // Prompt and wait for user command input.
  console.log("\nWelcome to the Cloud Mail Gmail Demo.");
  console.log("Follow the prompts to view and send emails using the component.");
  console.log("------------------------------------------------------------\n");
  displayMenu();
  menuPrompt();
  
  // Handle incoming commands.
  rl.on("line", async (line) => {
    // Handle follow-up prompts.
    switch (lastPrompt) {
      case "chooseMessage":
        try {
          isListingEmails = false;
          gmail1.config("ResponseType = 0");
          await gmail1.retrieveMessageRaw(line.trim());
          console.log(gmail1.getMessage());
          lastPrompt = "";
        } catch(e) {
          displayError(e);
        }
      break;
      case "sendTo":
        if (line !== "") {
          gmail1.setMessageTo(line.trim());
          prompt("cc", "CC", ":", "");
        } else {
          console.log("\nMust specify at least one email recipient.");
          lastPrompt = "";
        }
      break;
      case "cc":
        gmail1.setMessageCc(line.trim());
        prompt("bcc", "BCC", ":", "");
      break;
      case "bcc":
        gmail1.setMessageBcc(line.trim());
        prompt("subject", "Subject", ":", "");
      break;
      case "subject":
        gmail1.setMessageSubject(line.trim());
        prompt("messageBodyType", "Message Body Type (text/html)", ":", "");
      break;
      case "messageBodyType":
        if (line.toLowerCase() === "text" || line.toLowerCase() === "html") {
          gmail1.setMessageBodyContentType(line);
          prompt("content", "Message Content", ":", "");
        } else {
          console.log("\nPlease enter either text or html as a value.");
          prompt("messageBodyType", "MessageBodyType (text/html)", ":", "");
        }
      break;
      case "content":
        gmail1.setMessageBodyContent(line);
        prompt("attachment", "Add Attachment(s) (separated by semicolons)", ":", "");
      break;
      case "attachment":
        gmail1.setMessageAttachments(line.trim());
        console.log("\nSending email...");
        await gmail1.sendMail();
        console.log("\nEmail sent!");
        lastPrompt = "";
      break;
    }

    // Handle menu commands.
    switch (line) {
      case "l":
        isListingEmails = true;
        gmail1.config("ResponseType = 4");
        await gmail1.listMessages("", "");
        menuPrompt();
      break;
      case "v":
        prompt("chooseMessage", "Message to view", ":", "");
      break;
      case "c":
        prompt("sendTo", "Send to", ":", "");
      break;
      case "?":
        displayMenu();
        menuPrompt();
      break;
      case "q":
        process.exit(0);
      default:
        if (lastPrompt === "") {
          displayMenu();
          menuPrompt();
        }
      break;
    }
  });
}

function displayError(e) {
  process.stdout.write("Error");
  if (e instanceof cloudmail.gmail) {
    console.error(` (${e.code})`);
  }
  console.error(`: ${e.message}`);
  console.trace();
  process.exit(1);
}

function displayMenu() {
  console.log("\nGMail Commands:");
  console.log("l                   list next 100 messages");
  console.log("v                   view the content of choice message");
  console.log("c                   compose a message");
  console.log("?                   display menu options");
  console.log("q                   quit\n");
}

function menuPrompt() {
  process.stdout.write("gmail> ");
  lastPrompt = "";
}



function prompt(promptName, label, punctuation, defaultVal)
{
  lastPrompt = promptName;
  lastDefault = defaultVal;
  process.stdout.write(`${label} [${defaultVal}]${punctuation} `);
}
